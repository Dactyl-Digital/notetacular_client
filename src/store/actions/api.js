export const API_REQUEST = "API_REQUEST"

// Example usage:
// apiRequest({ url: "phoenix_api/list-toppins", onSuccess, onError });

// NOTE: Response from api subject to change...
// const onSuccess = apiResponse => {
//   const { data: { toppings } } = apiResponse;
//   listToppingsAction(toppings);
// };

// const onError = apiResponse => {
//   const { data: { message } } = apiResponse;
//   showToppingsError(message);
// };

export const apiRequest = ({
  method = "GET",
  url,
  parentResource,
  payload = {},
  loadingResource,
  onSuccess,
  onError,
}) => ({
  type: API_REQUEST,
  payload,
  loadingResource,
  meta: { method, url, parentResource, onSuccess, onError },
})
