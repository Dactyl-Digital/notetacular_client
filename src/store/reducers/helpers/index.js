export const checkProperty = ({
  obj,
  property,
  successFn,
  failFn,
  recursiveFn,
}) =>
  obj.hasOwnProperty(property)
    ? typeof recursiveFn === "function"
      ? recursiveFn(obj[property])
      : successFn()
    : failFn()
