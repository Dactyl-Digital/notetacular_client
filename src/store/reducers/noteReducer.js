import {
  SET_CREATED_NOTE,
  SET_NOTE_LIST,
  LIST_SHARED_NOTES,
  SET_CREATE_NOTE_ERROR,
  SET_NOTE_LIST_ERROR,
  SET_LIST_SHARED_NOTE_ERROR,
} from "../actions/note"
// import {helperFunction} from '../helpers'

export const noteInitialState = {
  listNotesOffset: 0,
  notes: {},
  listSharedNotesOffset: 0,
  sharedNotes: {},
  createNoteError: null,
  noteListError: null,
  listSharedNotesError: null,
}

const normalizeSingle = (noteState, { data }) => ({
  [data.topic_id]: {
    ...noteState.notes[data.topic_id],
    [data.id]: data,
  },
})

// TODO: Move this into a helper folder.
const normalize = key => ({ data }) =>
  data[key].reduce((acc, resource) => {
    if (acc[resource.topic_id]) {
      // Updating a current topic's notes
      acc = {
        ...acc,
        [resource.topic_id]: {
          ...acc[resource.topic_id],
          [resource.id]: resource,
        },
      }
    } else {
      // Creating a new entry for a topic's notes
      acc[resource.topic_id] = {
        [resource.id]: resource,
      }
    }
    return acc
  }, {})

const normalizeNotes = normalize("notes")

export default function noteReducer(
  noteState = noteInitialState,
  { type, payload }
) {
  if (type === SET_CREATED_NOTE) {
    console.log("Da payload in create note is:")
    console.log(payload)
    return {
      ...noteState,
      notes: {
        ...noteState.notes,
        ...normalizeSingle(noteState, payload),
      },
    }
  }
  if (type === SET_NOTE_LIST) {
    return noteListNewState(noteState, payload)
  }
  // if (type === LIST_SHARED_SUB_CATEGORIES) {
  //   return { ...noteState, successfulSignup: true }
  // }
  if (type === SET_CREATE_NOTE_ERROR) {
    return { ...noteState, createNoteError: payload }
  }
  if (type === SET_NOTE_LIST_ERROR) {
    return { ...noteState, noteListError: payload }
  }
  // if (type === SET_LIST_SHARED_NOTES_ERROR) {
  //   return { ...noteState, signinError: payload.errors }
  // }
  return noteState
}

const noteListNewState = (noteState, payload) => {
  console.log("the payload in list new note state:")
  console.log(payload)
  return {
    ...noteState,
    notes: {
      ...noteState.notes,
      ...normalizeNotes(payload),
    },
    // NOTE: w/ the new state shape of:
    // {
    //   [topic1_id]: {
    //     [note1_id]: {},
    //     [note2_id]: {},
    //   },
    //   [topic2_id]: {
    //     [note3_id]: {},
    //     [note4_id]: {},
    //   },
    // }
    // You'll need to manage an offset for each individual topic's notes...
    // listNotesOffset: noteState.listNotesOffset + 20,
  }
}
