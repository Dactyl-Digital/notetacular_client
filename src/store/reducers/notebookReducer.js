import {
  SET_CREATED_NOTEBOOK,
  SET_NOTEBOOK_LIST,
  LIST_SHARED_NOTEBOOKS,
  SET_CREATE_NOTEBOOK_ERROR,
  SET_NOTEBOOK_LIST_ERROR,
  SET_LIST_SHARED_NOTEBOOKS_ERROR,
} from "../actions/notebook"
// import {helperFunction} from '../helpers'

const listNotebooksOffset = JSON.parse(
  localStorage.getItem("listNotebooksOffset")
)
const notebooks = JSON.parse(localStorage.getItem("notebooks"))
const notebooksPaginationEnd = JSON.parse(
  localStorage.getItem("notebooksPaginationEnd")
)

export const notebookInitialState = {
  listNotebooksOffset: listNotebooksOffset ? listNotebooksOffset : 0,
  notebooks: notebooks ? notebooks : {},
  notebooksPaginationEnd: notebooksPaginationEnd
    ? notebooksPaginationEnd
    : false,
  listSharedNotebooksOffset: 0,
  sharedNotebooks: {},
  createNotebookError: null,
  notebookListError: null,
  listSharedNotebooksError: null,
}

const normalizeSingle = ({ data }) => ({
  [data.id]: {
    ...data,
  },
})

// TODO: Move this into a helper folder.
const normalize = key => ({ data }) =>
  data[key].reduce((acc, resource) => {
    acc[resource.id] = resource
    return acc
  }, {})

const normalizeNotebooks = normalize("notebooks")

export default function notebookReducer(
  notebookState = notebookInitialState,
  { type, payload }
) {
  if (type === SET_CREATED_NOTEBOOK) {
    const result = {
      ...notebookState,
      notebooks: { ...notebookState.notebooks, ...normalizeSingle(payload) },
    }

    return result
  }
  if (type === SET_NOTEBOOK_LIST) {
    return notebookListNewState(notebookState, payload)
  }
  // if (type === LIST_SHARED_NOTEBOOKS) {
  //   return { ...notebookState, successfulSignup: true }
  // }
  if (type === SET_CREATE_NOTEBOOK_ERROR) {
    return { ...notebookState, createNotebookError: payload }
  }
  if (type === SET_NOTEBOOK_LIST_ERROR) {
    return { ...notebookState, notebookListError: payload }
  }
  // if (type === SET_LIST_SHARED_NOTEBOOKS_ERROR) {
  //   return { ...notebookState, signinError: payload.errors }
  // }
  return notebookState
}

const notebookListNewState = (notebookState, payload) => {
  const notebooksPaginationEnd = payload.data.notebooks.length !== 20
  return {
    ...notebookState,
    notebooks: {
      ...notebookState.notebooks,
      ...normalizeNotebooks(payload),
    },
    listNotebooksOffset: notebookState.listNotebooksOffset + 20,
    notebooksPaginationEnd,
  }
}
