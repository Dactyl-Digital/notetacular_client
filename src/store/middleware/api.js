import axios from "axios"
import { toggleLoader } from "../actions/ui"

const API_REQUEST = "API_REQUEST"
const VALID_SESSION = "VALID_SESSION"

const invalidSession = ({ response: { data } }) => {
  console.log("invalidSession running")
  if (data.hasOwnProperty("message") && data.message === "Invalid session") {
    return {
      type: "INVALID_SESSION",
      payload: {},
      meta: {
        trigger: "User's cookie session has expired.",
      },
    }
  }
  return VALID_SESSION
}

const dispatchToggleLoader = (dispatch, bool, { method, url }) =>
  dispatch(toggleLoader({ loading: bool, trigger: `${method} ${url}` }))

const makeRequest = (
  dispatch,
  { payload },
  { method, url, onSuccess, onError }
) => {
  dispatchToggleLoader(dispatch, true, { method, url })
  if (method === "GET") {
    axios
      .get(url, {
        params: payload,
      })
      .then(function(response) {
        dispatchToggleLoader(dispatch, false, { method, url })
        onSuccess(response)
      })
      .catch(function(error) {
        console.log("the error in GET")
        console.log(error)
        if (
          error.hasOwnProperty("message") &&
          error.message === "Network Error"
        ) {
          return onError({
            response: {
              data: { message: "An error occured with the network." },
            },
          })
        }
        dispatchToggleLoader(dispatch, false, { method, url })
        const result = invalidSession(error)
        if (result === VALID_SESSION) return onError(error)
        dispatch(result)
      })
  }
  if (method === "POST") {
    // TODO: might break these out into individual functions as
    // postRequest, getRequest, etc... if generalizing it
    // via the method turns out to be a pain.
    // dispatchToggleLoader(dispatch, true, { method, url })
    return axios
      .post(url, payload)
      .then(function(response) {
        dispatchToggleLoader(dispatch, false, { method, url })
        onSuccess(response)
      })
      .catch(function(error) {
        // DOC:
        // Status Code 400 Err Response comes back as:
        // error.response.data.errors = {
        //   username: ["That username is already taken"]
        // }
        // Status Code 500 Err Response comes back as:
        // error.response.data = {
        //   message: "Oops... Something went wrong."
        // }
        console.log("the error in POST")
        console.dir(error)

        if (
          error.hasOwnProperty("message") &&
          error.message === "Network Error"
        ) {
          return onError({
            response: {
              data: { message: "An error occured with the network." },
            },
          })
        }
        dispatchToggleLoader(dispatch, false, { method, url })
        const result = invalidSession(error)
        if (result === VALID_SESSION) return onError(error)
        dispatch(result)
      })
  }
  if (method === "PUT") {
    axios
      .put(url, payload)
      .then(function(response) {
        dispatchToggleLoader(dispatch, false, { method, url })
        onSuccess(response)
      })
      .catch(function(error) {
        console.log("the error in PUT")
        console.log(error)
        if (
          error.hasOwnProperty("message") &&
          error.message === "Network Error"
        ) {
          return onError({
            response: {
              data: { message: "An error occured with the network." },
            },
          })
        }
        dispatchToggleLoader(dispatch, false, { method, url })
        const result = invalidSession(error)
        if (result === VALID_SESSION) return onError(error)
        dispatch(result)
      })
  }
  if (method === "PATCH") {
    axios
      .patch(url, payload)
      .then(function(response) {
        dispatchToggleLoader(dispatch, false, { method, url })
        onSuccess(response)
      })
      .catch(function(error) {
        console.log("the error in PATCH")
        console.log(error)
        if (
          error.hasOwnProperty("message") &&
          error.message === "Network Error"
        ) {
          return onError({
            response: {
              data: { message: "An error occured with the network." },
            },
          })
        }
        dispatchToggleLoader(dispatch, false, { method, url })
        const result = invalidSession(error)
        if (result === VALID_SESSION) return onError(error)
        dispatch(result)
      })
  }
  if (method === "DELETE") {
    axios
      .delete(url)
      .then(function(response) {
        dispatchToggleLoader(dispatch, false, { method, url })
        onSuccess(response)
      })
      .catch(function(error) {
        console.log("the error in DELETE")
        console.log(error)
        if (
          error.hasOwnProperty("message") &&
          error.message === "Network Error"
        ) {
          return onError({
            response: {
              data: { message: "An error occured with the network." },
            },
          })
        }
        dispatchToggleLoader(dispatch, false, { method, url })
        const result = invalidSession(error)
        if (result === VALID_SESSION) return onError(error)
        dispatch(result)
      })
  }
}

export const apiMiddleware = ({ dispatch }) => next => action => {
  console.log("Inside the middleware w/ action: ", action)
  if (action.type === API_REQUEST) {
    console.log("action.type === API_REQUEST")
    return makeRequest(dispatch, action, action.meta)
  }
  console.log("calling next(action) in middleware")
  next(action)
}
