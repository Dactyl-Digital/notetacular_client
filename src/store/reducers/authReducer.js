import { LOGIN_USER } from "../actions/auth"
// import {helperFunction} from '../helpers'

const initialState = {
  authenticated: false,
  signupUserError: null,
  signinUserError: null,
}

export default function authReducer(
  authState = initialState,
  { type, payload }
) {
  if (type === LOGIN_USER) {
    return { ...authState, authenticated: true }
  }
  return authState
}
