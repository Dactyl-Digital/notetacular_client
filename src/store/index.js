import { createStore, applyMiddleware, compose } from "redux"
import { apiMiddleware } from "./middleware/api"
import { rootReducer } from "./reducers"

const composeEnhancers =
  (typeof window !== "undefined" && process.env.NODE_ENV === "development" &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose

export const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(apiMiddleware))
)

const manageAuthenticatedStatus = (lsAuthenticated, currentAuthenticated) => {
  // Previous auth session not saved in localStorage, but user is currently authenticated... so persist it.
  if (!lsAuthenticated && currentAuthenticated) {
    return typeof localStorage !== "undefined"
      ? localStorage.setItem("authenticated", currentAuthenticated)
      : null
  }
  // User has logged out. previous auth session saved in localStorage, and redux store
  // authenticated state has transitioned from true to false. So clear the localStorage
  if (!currentAuthenticated && lsAuthenticated) {
    return typeof localStorage !== "undefined"
      ? localStorage.removeItem("authenticated")
      : null
  }
  // FINAL CASE:
  // User not authenticated, but previous auth session was saved in localStorage, so restore it.
  // This case is handled in authReducer.
}

let currentAuthenticated
const handleChange = () => {
  let previousAuthenticated = currentAuthenticated
  const {
    auth: { authenticated: currentAuthenticated },
  } = store.getState()
  const lsAuthenticated =
    typeof localStorage !== "undefined"
      ? localStorage.getItem("authenticated")
      : null
  if (previousAuthenticated !== currentAuthenticated) {
    manageAuthenticatedStatus(lsAuthenticated, currentAuthenticated)
  }
}

store.subscribe(handleChange)
