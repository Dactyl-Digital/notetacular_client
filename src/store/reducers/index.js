import { combineReducers } from "redux"
// import uiReducer from "./uiReducer";
import authReducer from "./authReducer"
import notebookReducer from "./notebookReducer"
import subCategoryReducer from "./subCategoryReducer"
import topicReducer from "./topicReducer"
import noteReducer from "./noteReducer"
import noteTimerReducer from "./noteTimerReducer"
// import modalReducer from "./modalReducer"

export const rootReducer = combineReducers({
  // ui: uiReducer,
  auth: authReducer,
  notebook: notebookReducer,
  subCategory: subCategoryReducer,
  topic: topicReducer,
  note: noteReducer,
  noteTimer: noteTimerReducer,
  // modal: modalReducer,
})
