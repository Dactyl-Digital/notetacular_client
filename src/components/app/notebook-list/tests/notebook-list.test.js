import React from "react"
import { createStore, applyMiddleware } from "redux"
import axios from "axios"
import configureStore from "redux-mock-store"
import { rootReducer } from "../../../../store/reducers"
import { apiMiddleware } from "../../../../store/middleware/api"
import { Provider } from "react-redux"
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
  debugDOM,
} from "@testing-library/react"
// import { store } from "../../../../store"
import NotebookList from "../index"
// import { initialStateWithoutTransactions } from "test_fixture_data"
// import { inputChangeTestCase, inputChange } from "./helpers"

// this adds custom jest matchers from jest-dom
import "@testing-library/jest-dom/extend-expect"

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

jest.mock("axios")

test("NotebookList renders list of notebooks", async () => {
  // const store = mockStore(rootReducer)
  const store = createStore(rootReducer, {}, applyMiddleware(apiMiddleware))
  const { getByTestId, debug } = render(
    <Provider store={store}>
      <NotebookList />
    </Provider>
  )

  const firstNotebook = await waitForElement(() =>
    getByTestId("notebook-list-page")
  )
  debug()
  expect(firstNotebook).toMatchSnapshot()
})

// test("BudgetDisplay renders with retrieved data from budget app API.", async done => {
//   const { getByTestId, debug } = render(
//     <BudgetReducerProvider
//       reducer={{ state: initialStateWithoutTransactions, dispatch: () => true }}
//     >
//       <BudgetDisplay budgetData={initialStateWithoutTransactions.data} />
//     </BudgetReducerProvider>
//   )
//   // debug()
//   const accountBalance = await waitForElement(() =>
//     getByTestId("account-balance")
//   )
//   const currentBudget = await waitForElement(() =>
//     getByTestId("current-budget")
//   )
//   const budgetExceeded = await waitForElement(() =>
//     getByTestId("budget-exceeded")
//   )
//   expect(accountBalance).toHaveTextContent("0")
//   expect(currentBudget).toHaveTextContent("Set Budget")
//   expect(budgetExceeded).toHaveTextContent("You're Good")
//   done()
// })
