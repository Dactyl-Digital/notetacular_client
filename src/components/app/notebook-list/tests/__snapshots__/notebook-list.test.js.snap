// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotebookList renders list of notebooks 1`] = `
<div
  class="sc-jzJRlG gATEGt"
  data-testid="notebook-list-page"
>
  <div
    class="sc-EHOje guJqFN"
  >
    <nav>
      <div
        class="sc-htpNat LvVay"
      >
        <span>
          DDArrow
        </span>
        <span
          id="more-options"
        >
          More Options
        </span>
      </div>
      <nav
        class="sc-bxivhb fjvDEI"
      >
        <a
          class="sc-ifAKCX frQXxE scroll-link"
          href="#Notebook1"
        >
          <div
            class="circle"
          />
          <span>
            Notebook1
          </span>
        </a>
        <a
          class="sc-ifAKCX bjFYnq scroll-link"
          href="#Notebook2"
        >
          <div
            class="circle circle-last-child"
          />
          <span>
            Notebook2
          </span>
        </a>
      </nav>
      <button>
        Log Out
      </button>
    </nav>
  </div>
  <div
    id="main-content"
  >
    <div
      class="sc-bwzfXH jCcQAv"
    >
      <h1>
        Notebooks
      </h1>
      <svg
        class="sc-bdVaJa giaiHc"
        id="Capa_1"
        version="1.1"
        viewBox="0 0 310.42 310.42"
        x="0"
        xmlns="http://www.w3.org/2000/svg"
        y="0"
      >
        <g>
          <g>
            <path
              d="M273.587 214.965c49.11-49.111 49.109-129.021 0-178.132-49.111-49.111-129.02-49.111-178.13 0C53.793 78.497 47.483 140.462 76.51 188.85c0 0 2.085 3.498-.731 6.312l-64.263 64.263c-12.791 12.79-15.836 30.675-4.493 42.02l1.953 1.951c11.343 11.345 29.229 8.301 42.019-4.49l64.128-64.128c2.951-2.951 6.448-.866 6.448-.866 48.387 29.026 110.352 22.717 152.016-18.947zM118.711 191.71c-36.288-36.288-36.287-95.332.001-131.62 36.288-36.287 95.332-36.288 131.619 0 36.288 36.287 36.288 95.332 0 131.62-36.288 36.286-95.331 36.286-131.62 0z"
            />
            <g>
              <path
                d="M126.75 118.424c-1.689 0-3.406-.332-5.061-1.031-6.611-2.798-9.704-10.426-6.906-17.038 17.586-41.559 65.703-61.062 107.261-43.476 6.611 2.798 9.704 10.426 6.906 17.038-2.799 6.612-10.425 9.703-17.039 6.906-28.354-11.998-61.186 1.309-73.183 29.663-2.099 4.959-6.913 7.938-11.978 7.938z"
              />
            </g>
          </g>
        </g>
      </svg>
    </div>
    <div
      class="sc-htoDjs kStOlp"
      id="modal"
    />
    <div
      id="notebook-list"
    >
      <div
        class="sc-fjdhpX fiEohR"
        id="Notebook1"
      >
        <div
          id="title-and-tags"
        >
          <a
            href="/app/notebook/1/sub-categories"
          >
            <div
              class="title-container"
            >
              <h3>
                Notebook1
              </h3>
            </div>
          </a>
        </div>
        <div
          id="icons"
        >
          <svg
            class="sc-bdVaJa giaiHc"
            id="Layer_1"
            version="1.1"
            viewBox="0 0 512 512"
            x="0"
            xmlns="http://www.w3.org/2000/svg"
            y="0"
          >
            <g>
              <g>
                <path
                  d="M436 60h-75V45c0-24.813-20.187-45-45-45H196c-24.813 0-45 20.187-45 45v15H76c-24.813 0-45 20.187-45 45 0 19.928 13.025 36.861 31.005 42.761L88.76 470.736C90.687 493.875 110.385 512 133.604 512h244.792c23.22 0 42.918-18.125 44.846-41.271l26.753-322.969C467.975 141.861 481 124.928 481 105c0-24.813-20.187-45-45-45zM181 45c0-8.271 6.729-15 15-15h120c8.271 0 15 6.729 15 15v15H181V45zm212.344 423.246c-.643 7.712-7.208 13.754-14.948 13.754H133.604c-7.739 0-14.305-6.042-14.946-13.747L92.294 150h327.412l-26.362 318.246zM436 120H76c-8.271 0-15-6.729-15-15s6.729-15 15-15h360c8.271 0 15 6.729 15 15s-6.729 15-15 15z"
                />
              </g>
            </g>
            <g>
              <g>
                <path
                  d="M195.971 436.071l-15-242c-.513-8.269-7.67-14.558-15.899-14.043-8.269.513-14.556 7.631-14.044 15.899l15 242.001c.493 7.953 7.097 14.072 14.957 14.072 8.687 0 15.519-7.316 14.986-15.929z"
                />
              </g>
            </g>
            <g>
              <g>
                <path
                  d="M256 180c-8.284 0-15 6.716-15 15v242c0 8.284 6.716 15 15 15s15-6.716 15-15V195c0-8.284-6.716-15-15-15z"
                />
              </g>
            </g>
            <g>
              <g>
                <path
                  d="M346.927 180.029c-8.25-.513-15.387 5.774-15.899 14.043l-15 242c-.511 8.268 5.776 15.386 14.044 15.899 8.273.512 15.387-5.778 15.899-14.043l15-242c.512-8.269-5.775-15.387-14.044-15.899z"
                />
              </g>
            </g>
          </svg>
        </div>
      </div>
      <div
        class="sc-fjdhpX cCYHQd"
        id="Notebook2"
      >
        <div
          id="title-and-tags"
        >
          <a
            href="/app/notebook/2/sub-categories"
          >
            <div
              class="title-container"
            >
              <h3>
                Notebook2
              </h3>
            </div>
          </a>
        </div>
        <div
          id="icons"
        >
          <svg
            class="sc-bdVaJa giaiHc"
            id="Layer_1"
            version="1.1"
            viewBox="0 0 512 512"
            x="0"
            xmlns="http://www.w3.org/2000/svg"
            y="0"
          >
            <g>
              <g>
                <path
                  d="M436 60h-75V45c0-24.813-20.187-45-45-45H196c-24.813 0-45 20.187-45 45v15H76c-24.813 0-45 20.187-45 45 0 19.928 13.025 36.861 31.005 42.761L88.76 470.736C90.687 493.875 110.385 512 133.604 512h244.792c23.22 0 42.918-18.125 44.846-41.271l26.753-322.969C467.975 141.861 481 124.928 481 105c0-24.813-20.187-45-45-45zM181 45c0-8.271 6.729-15 15-15h120c8.271 0 15 6.729 15 15v15H181V45zm212.344 423.246c-.643 7.712-7.208 13.754-14.948 13.754H133.604c-7.739 0-14.305-6.042-14.946-13.747L92.294 150h327.412l-26.362 318.246zM436 120H76c-8.271 0-15-6.729-15-15s6.729-15 15-15h360c8.271 0 15 6.729 15 15s-6.729 15-15 15z"
                />
              </g>
            </g>
            <g>
              <g>
                <path
                  d="M195.971 436.071l-15-242c-.513-8.269-7.67-14.558-15.899-14.043-8.269.513-14.556 7.631-14.044 15.899l15 242.001c.493 7.953 7.097 14.072 14.957 14.072 8.687 0 15.519-7.316 14.986-15.929z"
                />
              </g>
            </g>
            <g>
              <g>
                <path
                  d="M256 180c-8.284 0-15 6.716-15 15v242c0 8.284 6.716 15 15 15s15-6.716 15-15V195c0-8.284-6.716-15-15-15z"
                />
              </g>
            </g>
            <g>
              <g>
                <path
                  d="M346.927 180.029c-8.25-.513-15.387 5.774-15.899 14.043l-15 242c-.511 8.268 5.776 15.386 14.044 15.899 8.273.512 15.387-5.778 15.899-14.043l15-242c.512-8.269-5.775-15.387-14.044-15.899z"
                />
              </g>
            </g>
          </svg>
        </div>
      </div>
    </div>
    <button>
      Load More
    </button>
  </div>
</div>
`;
